src/base/zonefile_fmt.rs

```diff
@@ -0,0 +1,337 @@
+use core::fmt;
+
+#[derive(Clone, Copy, Debug)]
+pub struct Error;
+
+impl From<fmt::Error> for Error {
+    fn from(_: fmt::Error) -> Self {
+        Self
+    }
+}
+
+pub type Result = core::result::Result<(), Error>;
+
+pub struct ZoneFileDisplay<'a, T: ?Sized> {
+    inner: &'a T,
+    pretty: bool,
+}
+
+impl<T: ZonefileFmt + ?Sized> fmt::Display for ZoneFileDisplay<'_, T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if self.pretty {
+            self.inner
+                .fmt(&mut MultiLineWriter::new(f))
+                .map_err(|_| fmt::Error)
+        } else {
+            self.inner
+                .fmt(&mut SimpleWriter::new(f))
+                .map_err(|_| fmt::Error)
+        }
+    }
+}
+
+/// Show a value as zonefile format
+pub trait ZonefileFmt {
+    fn fmt(&self, p: &mut impl Formatter) -> Result;
+
+    fn display_zonefile(&self, pretty: bool) -> ZoneFileDisplay<'_, Self> {
+        ZoneFileDisplay {
+            inner: self,
+            pretty,
+        }
+    }
+}
+
+impl<T: ZonefileFmt> ZonefileFmt for &T {
+    fn fmt(&self, p: &mut impl Formatter) -> Result {
+        T::fmt(self, p)
+    }
+}
+
+/// Determines how a zonefile is formatted
+pub trait FormatWriter: Sized {
+    /// Push a token to the zonefile
+    fn fmt_token(&mut self, args: fmt::Arguments<'_>) -> Result;
+
+    /// Start a block of grouped tokens
+    ///
+    /// This might push `'('` to the zonefile, but may be ignored by the
+    /// `PresentationWriter`.
+    fn begin_block(&mut self) -> Result;
+
+    /// End a block of grouped tokens
+    ///
+    /// This might push `'('` to the zonefile, but may be ignored by the
+    /// `PresentationWriter`.
+    fn end_block(&mut self) -> Result;
+
+    /// Write a comment
+    ///
+    /// This may be ignored.
+    fn fmt_comment(&mut self, args: fmt::Arguments<'_>) -> Result;
+
+    /// End the current record and start a new line
+    fn newline(&mut self) -> Result;
+}
+
+/// The simplest possible zonefile writer
+///
+/// This writer does not do any alignment, comments and squeezes each record
+/// onto a single line.
+struct SimpleWriter<'a> {
+    first: bool,
+    writer: &'a mut (dyn fmt::Write + 'a),
+}
```

Comment:
This could be defined as
```rust
struct SimpleWriter<W> {
    first: bool,
    writer: W,
}
```
The primary problem is that `fmt::Formatter` already contains a trait object, and using it in this struct will result in two layers of vtables.